### ISSUE TITLE
CJS doesn't support recursion

### ISSUE THREAD

**Issue by dchambers** [dchambers/composerjs#21]
At present, ComposerJs doesn't support recursive data structures. On the face of it this could be fixed by adding an `addOptionalNode()` method that looks like this:

``` js
node.addOptionalNode('child', node);
```

Which allows a node to optionally have other nodes of a particular type. To make this more powerful, type specialization could be extended to also work on regular nodes, in addition to node-lists as at present, for example:

``` js
node('typeA').set('x', true);
node('typeB').set('y', true);
node.addOptionalNode('child', node);
```

However, this creates a problem because handlers are currently specified globally, rather than relative to the node in question. For example, given a model that looks like this:

``` js
node.addNodeList('nodes');
node.addOptionalNode('childNode', node);
node.nodes.set('amount', 1);
node.addHandler([node.nodes.p('amount')], [node.p('total')], function(in, out) {
    out.total = in.sum();
});
```

this would appear to cause `node.total` to be equal to the sum of the child `amount` properties, but in fact it would actually be a grand total of all child `amount` properties, for all nodes.

To fix this, we would need to start requiring that all handlers are specified relative to the node the handler is being added to, so that:

``` js
node.addHandler([model.p('x')], [node.p('y')], function(in, out, current, index) {
    // ...
});
```

would instead need to be written as:

``` js
node.addHandler([node.parent().p('x')], [node.p('y')], function(in, out, current, index) {
    // ...
});
```

However, in addition to these three changes, recursion would complicate the [main handler-loop](https://github.com/dchambers/composerjs/issues/15#issuecomment-71443614), as it would no longer be possible to process the DAG only when the model is sealed.

**Comment by dchambers** [comment]
@james-shaw-turner, does this make sense?

**Comment by dchambers** [comment]
On second thoughts, the DAG problem can be solved by recalculating the ordered handler list each time an optional node is added or removed. This would require distinct instances of the same handlers to be created internally, in much the same way as if the 'recursion' had been set-up manually, prior to sealing.

For example, consider the following code which effectively creates a _pre-seal_ recursive node tree:

``` js
model.set('enabled', true);
model.addNode('child');
buildChild(model, 3);

function buildChild(parent, index) {
    var node = parent.child;
    node.addHandler([parent.p('enabled')], [node.p('enabled')], function(in, out) {
        out.enabled = in.enabled;
    });

    if(index > 0) {
        node.addNode('child');
        buildChild(node, index--);
    }
}
```

As far as ComposerJs is concerned, this code creates a model with 3 distinct handlers, each operating on distinct input and output properties.

In this way, recursion can be supported with the following changes:
1. A new `addOptionalNode()` method would be needed that allows a previously defined node to be referred to.
2. `addHandler()` would need to be changed to enforce that node properties were always specified relative to the node being added (using `node.parent()` and `node.root()`), where this relative information was reflected in the final _property-specifier_ created.
3. Property specifiers would need to be returned even when `as()` isn't invoked, as they would need to contain the relative path information.
4. The DAG would need to be re-created each time an optional node was added or removed.
5. We could optionally choose to allow nodes to be specialized in the same way that node-lists currently can.

@james-shaw-turner, I believe this fixes the last major risk with ComposerJs. I'll update the docs to reflect these changes in due course.

**Comment by dchambers** [comment]
I've now updated the list of changes necessary to support recursion, some which have already been done:
1. The `hasBeenUpdated()` method has been renamed to `markAsUpdated()`, but has been retained ([see reasons](https://github.com/dchambers/composerjs/issues/22)).
2. Handler objects have been dropped, and replaced by [stateful handlers](https://github.com/dchambers/composerjs#stateful-handlers).
3. Property specifiers have been changed so that they are [always given relatively](https://github.com/dchambers/composerjs#property-specifiers), and they have also been clarified in the docs so they are explained in detail.
4. The `addNode()` method has been renamed to `addOptionalNode()`, and there are now `create()` and `delete()` companion methods for dynamically creating these nodes. (**yet to be done**).
5. `addOptionalNode()` has been upgraded so that it supports specializations in the same way as `addNodeList()` (**yet to be done**).
